analyzers/src/SonarAnalyzer.CSharp/Rules/VariableShadowsField.cs

```diff
@@ -28,75 +28,94 @@ public sealed class VariableShadowsField : SonarDiagnosticAnalyzer
+                _ when DeclarationPatternSyntaxWrapper.IsInstance(node) => ((DeclarationPatternSyntaxWrapper)node).Designation,
+                _ when ListPatternSyntaxWrapper.IsInstance(node) => ((ListPatternSyntaxWrapper)node).Designation,
+                _ => null
+            };
 
-        private static void ProcessVariableDesignation(SonarSyntaxNodeReportingContext context, VariableDesignationSyntaxWrapper variableDesignation)
+        private static void Process(SonarSyntaxNodeReportingContext context, SyntaxNode node)
         {
-            if (variableDesignation.AllVariables() is { Length: > 0 } variables
-                && context.ContainingSymbol.ContainingType is { } containingType
-                && GetMembers(containingType) is var members)
+            if (ExtractIdentifiers(node) is { Count: > 0 } identifiers
+                && GetContextSymbols(context) is var members)
             {
-                foreach (var variable in variables)
+                foreach (var identifier in identifiers)
                 {
-                    ReportOnVariableMatchingField(context, members, variable.Identifier);
+                    ReportOnVariableMatchingField(context, members, identifier);
                 }
             }
         }
 
-        private static void ProcessVariableDeclaration(SonarSyntaxNodeReportingContext context, VariableDeclarationSyntax variableDeclaration)
-        {
-            if (variableDeclaration is { Variables: { Count: > 0 } variables }
-                && context.ContainingSymbol.ContainingType is { } containingType
-                && GetMembers(containingType) is var members)
+        private static List<SyntaxToken> ExtractIdentifiers(SyntaxNode node) =>
+            node switch
             {
-                foreach (var variable in variables)
-                {
-                    ReportOnVariableMatchingField(context, members, variable.Identifier);
-                }
-            }
+                VariableDeclarationSyntax variableDeclaration => variableDeclaration.Variables.Select(x => x.Identifier).ToList(),
+                ForEachStatementSyntax foreachStatement => new() { foreachStatement.Identifier },
+                _ when VariableDesignationSyntaxWrapper.IsInstance(node) => ((VariableDesignationSyntaxWrapper)node).AllVariables().Select(x => x.Identifier).ToList(),
+                _ => new()
+            };
+
+        private static List<ISymbol> GetContextSymbols(SonarSyntaxNodeReportingContext context)
+        {
+            var symbols = ((INamespaceOrTypeSymbol)context.ContainingSymbol.ContainingSymbol).GetMembers();
+            var primaryCtor = symbols.FirstOrDefault(x => x is IMethodSymbol && IsPrimaryCtor(x));
+            var relevantSymbols = symbols.Where(x => x is IPropertySymbol or IFieldSymbol).ToList();
+            return primaryCtor is null ? relevantSymbols : relevantSymbols.Concat(primaryCtor.GetParameters()).ToList();
```

Comment:
Some naming suggestions:
```suggestion
            var members = ((INamespaceOrTypeSymbol)context.ContainingSymbol.ContainingSymbol).GetMembers();
            var primaryConstructorParameter = symbols.FirstOrDefault(x => x is IMethodSymbol && IsPrimaryCtor(x))?.GetParameters();
            var fieldsAndProperties = symbols.Where(x => x is IPropertySymbol or IFieldSymbol).ToList();
            return primaryCtor is null ? relevantSymbols : relevantSymbols.Concat(primaryCtor.GetParameters()).ToList();
```